crates/wasi/src/preview2/host/io.rs

```diff
@@ -256,85 +207,129 @@ impl<T: WasiView> streams::Host for T {
-        let s = self.table_mut().get_internal_output_stream_mut(stream)?;
-        loop {
-            if let InternalOutputStream::Host(s) = s {
-                HostOutputStream::ready(s.as_mut()).await?;
-            }
-            let mut bytes = bytes::Bytes::from_static(ZEROS);
-            bytes.truncate(remaining.min(bytes.len()));
-            let (written, state) = match s {
-                InternalOutputStream::Host(s) => match HostOutputStream::write(s.as_mut(), bytes) {
-                    Ok(a) => a,
-                    Err(e) => {
-                        if let Some(e) = e.downcast_ref::<StreamRuntimeError>() {
-                            tracing::debug!("stream runtime error: {e:?}");
-                            return Ok(Err(()));
-                        } else {
-                            return Err(e);
-                        }
-                    }
-                },
-                InternalOutputStream::File(s) => match FileOutputStream::write(s, bytes).await {
-                    Ok(a) => a,
-                    Err(e) => {
-                        if let Some(e) = e.downcast_ref::<StreamRuntimeError>() {
-                            tracing::debug!("stream runtime error: {e:?}");
-                            return Ok(Err(()));
-                        } else {
-                            return Err(e);
-                        }
-                    }
-                },
-            };
-            remaining -= written;
-            if remaining == 0 || state == StreamState::Closed {
-                return Ok(Ok((len - remaining as u64, state.into())));
-            }
+    ) -> anyhow::Result<Pollable> {
+        // Ensure that table element is an output-stream:
+        let _ = self.table_mut().get_output_stream_mut(stream)?;
+
+        fn output_stream_ready<'a>(stream: &'a mut dyn Any) -> PollableFuture<'a> {
+            let stream = stream
+                .downcast_mut::<Box<dyn HostOutputStream>>()
+                .expect("downcast to HostOutputStream failed");
+            Box::pin(async move {
+                let _ = stream.write_ready().await?;
+                Ok(())
+            })
         }
+
+        Ok(self
+            .table_mut()
+            .push_host_pollable(HostPollable::TableEntry {
+                index: stream,
+                make_future: output_stream_ready,
+            })?)
     }
 
+    async fn blocking_write_and_flush(
+        &mut self,
+        stream: OutputStream,
+        bytes: Vec<u8>,
+    ) -> Result<(), streams::Error> {
+        let s = self.table_mut().get_output_stream_mut(stream)?;
+
+        if bytes.len() > 4096 {
+            return Err(streams::Error::trap(anyhow::anyhow!(
+                "Buffer too large for blocking-write-and-flush (expected at most 4096)"
+            )));
+        }
+
+        let mut bytes = bytes::Bytes::from(bytes);
+        while !bytes.is_empty() {
+            let permit = HostOutputStream::write_ready(s).await?;
```

Comment:
Syntax-wise, is it possible to write this as `s.write_read().await?`?

Reply:
fxied